File: core/src/main/java/com/gmail/marcosav2010/connection/ConnectionIdentificator.java
package com.gmail.marcosav2010.connection;

import com.gmail.marcosav2010.communicator.packet.packets.PacketIdentify;
import com.gmail.marcosav2010.peer.ConnectedPeer;

import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

/**
 * This clase manages all peer connections when connected peer registers
 *
 * @author Marcos
 */
public class ConnectionIdentificator extends NetworkIdentificator<ConnectedPeer> {

    private final Map<String, ConnectedPeer> namePeer;

    public ConnectionIdentificator() {
        namePeer = new ConcurrentHashMap<>();
    }

    protected ConnectedPeer identifyConnection(Connection connection, PacketIdentify info) {
        String name = getSuitableName(info.getName());
        ConnectedPeer c = new ConnectedPeer(name, info.getPeerUUID(), connection);
        namePeer.put(name, c);
        put(connection.getUUID(), c);
        return c;
    }

    private String getSuitableName(String providedName) {
        while (namePeer.containsKey(providedName))
            providedName += "_";
        return providedName;
    }

    public ConnectedPeer getPeer(String name) {
        return namePeer.get(name);
    }

    public boolean hasPeer(String name) {
        return namePeer.containsKey(name);
    }

    protected ConnectedPeer removePeer(UUID uuid) {
        ConnectedPeer c = remove(uuid);
        if (c == null)
            return null;

        return namePeer.remove(c.getName());
    }

    protected ConnectedPeer removePeer(String name) {
        ConnectedPeer c = namePeer.remove(name);
        if (c == null)
            return null;

        return remove(c.getConnection().getUUID());
    }
}

File: core/src/main/java/com/gmail/marcosav2010/connection/ConnectionIdentificator.kt
package com.gmail.marcosav2010.connection

import com.gmail.marcosav2010.communicator.packet.packets.PacketIdentify
import com.gmail.marcosav2010.peer.ConnectedPeer
import java.util.*
import java.util.concurrent.ConcurrentHashMap

/**
 * This class manages all peer connections when connected peer registers
 *
 * @author Marcos
 */
open class ConnectionIdentificator : NetworkIdentificator<ConnectedPeer>() {

    private val namePeer: MutableMap<String, ConnectedPeer> = ConcurrentHashMap()

    fun identifyConnection(connection: Connection, info: PacketIdentify): ConnectedPeer {
        val name = getSuitableName(info.name!!)
        val c = ConnectedPeer(name, info.peerUUID, connection)
        namePeer[name] = c
        put(connection.uuid!!, c)
        return c
    }

    private fun getSuitableName(providedName: String): String {
        var output = providedName
        while (namePeer.containsKey(providedName)) output += "_"
        return providedName
    }

    operator fun get(name: String) = namePeer[name]

    fun hasPeer(name: String) = namePeer.containsKey(name)

    fun removePeer(uuid: UUID): ConnectedPeer? {
        val c = remove(uuid) ?: return null
        return namePeer.remove(c.name)
    }

    protected fun removePeer(name: String): ConnectedPeer? {
        val c = namePeer.remove(name) ?: return null
        return remove(c.connection.uuid!!)
    }
}